internal/config/config.go

```diff
@@ -0,0 +1,153 @@
+	flagSet.IntVar(&cfg.ParallelUpgrades, "parallel", parallelDefault, parallelDescription)
+
+	for _, s := range []func() error{
+		func() error {
+			return validateLoginStatus(conn)
+		},
+		func() error {
+			return validateAPIVersion(conn)
+		},
+		func() error {
+			return read("access token", conn.AccessToken, &cfg.APIToken)
+		},
+		func() error {
+			return read("API endpoint", conn.ApiEndpoint, &cfg.APIEndpoint)
+		},
+		func() error {
+			return read("skip SSL validation", conn.IsSSLDisabled, &cfg.SkipSSLValidation)
+		},
+		func() error {
+			return flagSet.Parse(args)
+		},
+		func() error {
+			return validateParallelUpgrades(cfg.ParallelUpgrades)
+		},
+		func() (err error) {
+			cfg.BrokerName, err = readBrokerName(flagSet.Args())
+			return
+		},
+	} {
+		if err := s(); err != nil {
+			return Config{}, err
+		}
+	}
+	return cfg, nil
+}
+
+type Config struct {
+	BrokerName        string
+	APIToken          string
+	APIEndpoint       string
+	SkipSSLValidation bool
+	ParallelUpgrades  int
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . CLIConnection
+type CLIConnection interface {
+	IsLoggedIn() (bool, error)
+	AccessToken() (string, error)
+	ApiVersion() (string, error)
+	ApiEndpoint() (string, error)
+	IsSSLDisabled() (bool, error)
+}
+
+func Options() map[string]string {
+	return map[string]string{
+		fmt.Sprintf("-%s", parallelFlag): parallelDescription,
+	}
+}
+
+func read[T any](desc string, get func() (T, error), set *T) error {
+	data, err := get()
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", desc, err)
+	}
+
+	*set = data
+	return nil
+}
+
+func validateLoginStatus(conn CLIConnection) error {
+	loggedIn, err := conn.IsLoggedIn()
+	switch {
+	case err != nil:
+		return fmt.Errorf("error getting login status: %w", err)
+	case !loggedIn:
+		return fmt.Errorf("you must authenticate with the cf cli before running this command")
+	default:
+		return nil
+	}
+}
+
+func validateAPIVersion(conn CLIConnection) error {
+	ver, err := conn.ApiVersion()
+	if err != nil {
+		return fmt.Errorf("error retrieving API version: %w", err)
+	}
+
+	v, err := version.NewVersion(ver)
+	switch {
+	case err != nil:
+		return fmt.Errorf("error parsing API version: %w", err)
+	case v.GreaterThanOrEqual(version.Must(version.NewVersion("4"))):
+		return fmt.Errorf("plugin requires API major version v3, got: %q", v.String())
+	case v.LessThan(version.Must(version.NewVersion("3.99"))):
+		return fmt.Errorf("plugin requires minimum API version v3.99, got: %q", v.String())
+	default:
+		return nil
+	}
+}
+
+func validateParallelUpgrades(p int) error {
+	if p <= 0 || p > 100 {
+		printUsage()
+		return fmt.Errorf("number of parallel upgrades must be in the range of 1 to 100")
+	}
+	return nil
+}
+
+func readBrokerName(args []string) (string, error) {
+	switch len(args) {
+	case 0:
+		printUsage()
+		return "", fmt.Errorf("missing broker name")
+	case 1: // OK
+	default:
```

Comment:
I'm probably just missing something here, but won't this fail if a user specifies the `parallel` flag? In that case args would be ["brokerName", "-parallel=n"] so len == 2?

Reply:
`flagSet.Args()` should contain only the **unparsed** flags.

Reply:
There is an additional usability issue here where the flags need to be `-parallel 5 MyBrokerName` rather than `MyBrokerName -parallel 5`. This is because the flags parses stops parsing at the first non-flag argument. This might be expected for a command like `tar`, but we could improve on it with a further PR.